src/Numerics/Mesh/Metrics.jl

```diff
@@ -98,127 +100,298 @@ function creategrid!(x1, x2, x3, e2c, ξ1, ξ2, ξ3)
+"""
+function setup_geom_factors_data!(
+    vgeo::VolumeGeometry{NTuple{3,Int}, <:AbstractArray},
+    D1,
+    D2,
+    D3,
+)
+    Nq = (size(D1, 1), size(D2, 1), size(D3, 1))
+    nelem = div(length(vgeo.ωJ), prod(Nq))
+
+    vgeo.x1 = reshape(vgeo.x1, (Nq..., nelem))
+    vgeo.x2 = reshape(vgeo.x2, (Nq..., nelem))
+    vgeo.x3 = reshape(vgeo.x3, (Nq..., nelem))
+    vgeo.ωJ = reshape(vgeo.ωJ, (Nq..., nelem))
+    vgeo.JcV = reshape(vgeo.JcV, (Nq..., nelem))
+    vgeo.ξ1x1 = reshape(vgeo.ξ1x1, (Nq..., nelem))
+    vgeo.ξ2x1 = reshape(vgeo.ξ2x1, (Nq..., nelem))
+    vgeo.ξ3x1 = reshape(vgeo.ξ3x1, (Nq..., nelem))
+    vgeo.ξ1x2 = reshape(vgeo.ξ1x2, (Nq..., nelem))
+    vgeo.ξ2x2 = reshape(vgeo.ξ2x2, (Nq..., nelem))
+    vgeo.ξ3x2 = reshape(vgeo.ξ3x2, (Nq..., nelem))
+    vgeo.ξ1x3 = reshape(vgeo.ξ1x3, (Nq..., nelem))
+    vgeo.ξ2x3 = reshape(vgeo.ξ2x3, (Nq..., nelem))
+    vgeo.ξ3x3 = reshape(vgeo.ξ3x3, (Nq..., nelem))
+
+end
+
+"""
+    compute_dxdxi_jacobian!(vgeo, D1, D2)
+
+Input arguments:
+- vgeo::VolumeGeometry, a struct containing the volumetric geometric factors
+- D1::DAT2, 1-D derivative operator on the device in the first dimension
+- D2::DAT2, 1-D derivative operator on the device in the second dimension
+
+Compute the Jacobian matrix of the mapping from physical coordinates,
+`vgeo.x1`, `vgeo.x2` with respect to reference coordinates `ξ1`, `ξ2`,
+for each quadrature point in element e.
+"""
+function compute_dxdxi_jacobian!(
+    vgeo::VolumeGeometry{NTuple{2,Int}, <:AbstractArray},
+    D1,
+    D2
+)
+    T = eltype(vgeo.x1)
+    Nq = (size(D1, 1), size(D2, 1))
+
+    vgeo.x1ξ1 .= vgeo.x1ξ2 .= zero(T)
+    vgeo.x2ξ1 .= vgeo.x2ξ2 .= zero(T)
+
+    for e in 1:nelem
+        for j in 1:Nq[2], i in 1:Nq[1]
+            for n in 1:Nq[1]
+                vgeo.x1ξ1[i, j, e] += D1[i, n] * vgeo.x1[n, j, e]
+                vgeo.x2ξ1[i, j, e] += D1[i, n] * vgeo.x2[n, j, e]
+            end
+            for n in 1:Nq[2]
+                vgeo.x1ξ2[i, j, e] += D2[j, n] * vgeo.x1[i, n, e]
+                vgeo.x2ξ2[i, j, e] += D2[j, n] * vgeo.x2[i, n, e]
+            end
+        end
+    end
+
+    return vgeo
+end
+
+"""
+    compute_dxdxi_jacobian!(vgeo, D1, D2, D3)
+
+Input arguments:
+- vgeo::VolumeGeometry, a struct containing the volumetric geometric factors
+- D1::DAT2, 1-D derivative operator on the device in the first dimension
+- D2::DAT2, 1-D derivative operator on the device in the second dimension
+- D3::DAT2, 1-D derivative operator on the device in the third dimension
+
+Compute the Jacobian matrix of the mapping from physical coordinates,
+`vgeo.x1`, `vgeo.x2`, `vgeo.x3`, with respect to reference coordinates `ξ1`,
+`ξ2`, `ξ3`, for each quadrature point in element e.
+"""
+function compute_dxdxi_jacobian!(
+    vgeo::VolumeGeometry{NTuple{3,Int}, <:AbstractArray},
+    D1,
+    D2,
+    D3
+)
+
+    T = eltype(vgeo.x1)
+    Nq = (size(D1, 1), size(D2, 1), size(D3, 1))
+
```

Comment:
define local variables here:
```suggestion
r_x1 = reshape(vgeo.x1, (Nq..., nelem))
```
use a consistent convention so that e.g. `r_X` is `X` reshaped.

Reply:
I decided to call the local variables without the prefix. I thought it was clearer.